src/raw/iter.rs

```diff
@@ -49,56 +52,224 @@ impl<'a> Iter<'a> {
-                    num_bytes,
-                    self.doc.as_bytes().len() - start
-                ),
-            }));
+            return Err(Error::new_without_key(ErrorKind::new_malformed(format!(
+                "length exceeds remaining length of buffer: {} vs {}",
+                num_bytes,
+                self.doc.as_bytes().len() - start
+            ))));
         }
         Ok(())
     }
 
-    fn next_oid(&self, starting_at: usize) -> Result<ObjectId> {
-        self.verify_enough_bytes(starting_at, 12)?;
-        let oid = ObjectId::from_bytes(
-            self.doc.as_bytes()[starting_at..(starting_at + 12)]
-                .try_into()
-                .unwrap(), // ok because we know slice is 12 bytes long
-        );
-        Ok(oid)
-    }
-
-    fn next_document(&self, starting_at: usize) -> Result<&'a RawDocument> {
+    fn next_document_len(&self, starting_at: usize) -> Result<usize> {
         self.verify_enough_bytes(starting_at, MIN_BSON_DOCUMENT_SIZE as usize)?;
         let size = i32_from_slice(&self.doc.as_bytes()[starting_at..])? as usize;
 
         if size < MIN_BSON_DOCUMENT_SIZE as usize {
-            return Err(Error::new_without_key(ErrorKind::MalformedValue {
-                message: format!("document too small: {} bytes", size),
-            }));
+            return Err(Error::new_without_key(ErrorKind::new_malformed(format!(
+                "document too small: {} bytes",
+                size
+            ))));
         }
 
         self.verify_enough_bytes(starting_at, size)?;
-        let end = starting_at + size;
-
-        if self.doc.as_bytes()[end - 1] != 0 {
-            return Err(Error {
-                key: None,
-                kind: ErrorKind::MalformedValue {
-                    message: "not null terminated".into(),
-                },
-            });
+
+        if self.doc.as_bytes()[starting_at + size - 1] != 0 {
+            return Err(Error::new_without_key(ErrorKind::new_malformed(
+                "not null terminated",
+            )));
         }
-        RawDocument::from_bytes(&self.doc.as_bytes()[starting_at..end])
+        Ok(size)
+    }
+}
+
+#[derive(Clone)]
+pub struct RawElement<'a> {
+    key: &'a str,
+    kind: ElementType,
+    doc: &'a RawDocument,
+    start_at: usize,
+    size: usize,
+}
+
+impl<'a> TryInto<RawBsonRef<'a>> for RawElement<'a> {
+    type Error = Error;
+
+    fn try_into(self) -> Result<RawBsonRef<'a>> {
+        self.value()
+    }
+}
+
+impl<'a> TryInto<RawBson> for RawElement<'a> {
+    type Error = Error;
+
+    fn try_into(self) -> Result<RawBson> {
+        Ok(self.value()?.to_raw_bson())
+    }
+}
+
+impl<'a> TryInto<Bson> for RawElement<'a> {
+    type Error = Error;
+
+    fn try_into(self) -> Result<Bson> {
+        self.value()?.to_raw_bson().try_into()
+    }
+}
+
+impl<'a> RawElement<'a> {
+    pub fn len(&self) -> usize {
+        self.size
+    }
+
+    pub fn key(&self) -> String {
```

Comment:
Is there any particular reason to return the string as copy here rather than the ref?